### PR TITLE
Feature/stl parser

### PR DESCRIPTION
**Pull request by juliusnehring**
Implemented an STL-parser.
This was tested against Thingi10k and successfully parses all 9956 .stl files except `77942.stl`, which the authors of the dataset acknowledge as corrupt. 

The autodetect mode works correctly for all of them, even though the check for testing if a file is binary is pretty much just a heuristic.

### CODE REVIEW

File: src/babel-serializer/geometry/stl.cc [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+            t.v1.x = parse_float();
+            t.v1.y = parse_float();
+            t.v1.z = parse_float();
+            t.v2.x = parse_float();
+            t.v2.y = parse_float();
+            t.v2.z = parse_float();
+
+            // this is the "attribute_byte_count" but in practice is always a color
+            // see https://en.wikipedia.org/wiki/STL_(file_format)#Color%20in%20binary%20STL
+            auto const color = parse_uint16();
```
Comment (jkunstwald) [inline src/babel-serializer/geometry/stl.cc:125]: this info would also be useful in the geometry struct definition

Comment (juliusnehring) [reply src/babel-serializer/geometry/stl.cc:125]: It technically is, `We accept colors in VisCAM and SolidView, as well as Materialise Magics format`, however I can see that it is not obvious that this relates to color only.

Comment (jkunstwald) [reply src/babel-serializer/geometry/stl.cc:125]: oh nevermind, i though the color values themselves are still encoded in the output, but they're just intensities in [0, 255] right? in that case the type should probably be uint8

Comment (juliusnehring) [reply src/babel-serializer/geometry/stl.cc:125]: Yeah, I was unsure about the type. We could also do `tg::color<3, uint8>`, however I am unsure about the state of `tg::color` with non-float types.

Comment (jkunstwald) [reply src/babel-serializer/geometry/stl.cc:125]: the simple struct is fine i think, but byte signals "opaque memory" while it really is an integer

Comment (Philip-Trettner) [reply src/babel-serializer/geometry/stl.cc:125]: uint8_t is better for now. proper color management in tg is on my list though

File: src/babel-serializer/geometry/stl.cc [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+        {
+            on_error(data, data, "STL-file too short to be valid", severity::error);
+        }
+        if (!data_as_string_view.trim().starts_with("solid"))
+            return false;
+
+        // unfortunately many binary stl also start with "solid"
+        // also check for '\n' followed by "facet"
+        int idx = -1;
+        for (size_t i = 0; i < data_as_string_view.size(); ++i)
```
Comment (Philip-Trettner) [inline src/babel-serializer/geometry/stl.cc]: when https://github.com/project-arcana/clean-core/pull/105 is merged we have a `index_of` in string_view

File: src/babel-serializer/geometry/stl.cc [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+                geometry.triangle_color[i] = default_color;
+            }
+        }
+        if (pos != data.size())
+        {
+            on_error(data, data, "Binary STL-file contains excess data", severity::warning);
+        }
+    };
+
+    auto const parse_ascii = [&]() {
```
Comment (Philip-Trettner) [inline src/babel-serializer/geometry/stl.cc:154]: just yi: the () is not needed

File: src/babel-serializer/geometry/stl.cc (hunk 2 above)
Comment (juliusnehring) [reply src/babel-serializer/geometry/stl.cc]: Nice, that is a feature I was missing :+1:

File: src/babel-serializer/geometry/stl.hh [hunk 4]
```diff
@@ -1,3 +1,54 @@
...
+
+    struct color
+    {
+        std::byte r;
+        std::byte g;
+        std::byte b;
+        std::byte a;
+    };
+
+    /// The name is optional and always emtpy for binary STL files
```
Comment (Philip-Trettner) [inline src/babel-serializer/geometry/stl.hh]: typo :)

File: src/babel-serializer/geometry/stl.cc (hunk 2 above)
Comment (Philip-Trettner) [reply src/babel-serializer/geometry/stl.cc]: got merged :D

Comment (juliusnehring) [reply src/babel-serializer/geometry/stl.cc]: Updated :+1:

**Review by Philip-Trettner — APPROVED** [review]

**Pull request merged**